File: examples/16-websockets/python/ws-server/main.py
import asyncio
import json
import logging
import os
import uuid
from collections import defaultdict
from contextlib import asynccontextmanager

import redis.asyncio as redis
from fastapi import FastAPI, WebSocket, WebSocketDisconnect
from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("ws-server")

REDIS_URL = os.getenv("REDIS_URL", "redis://redis:6379")
OTEL_ENDPOINT = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT", "http://lgtm:4318")
POD_NAME = os.getenv("POD_NAME", f"ws-pod-{uuid.uuid4().hex[:6]}")

resource = Resource.create({"service.name": "ws-server", "pod.name": POD_NAME})
provider = TracerProvider(resource=resource)
provider.add_span_processor(BatchSpanProcessor(OTLPSpanExporter(endpoint=f"{OTEL_ENDPOINT}/v1/traces")))
trace.set_tracer_provider(provider)
tracer = trace.get_tracer("ws-server")

CHANNEL = "ws:broadcast"
clients: dict[str, WebSocket] = {}
seq_counters: dict[str, int] = defaultdict(int)
message_buffers: dict[str, list] = defaultdict(list)
BUFFER_SIZE = 100

redis_pool = None
pubsub_task = None


async def backplane_listener():
    sub = redis_pool.pubsub()
    await sub.subscribe(CHANNEL)
    async for msg in sub.listen():
        if msg["type"] != "message":
            continue
        payload = json.loads(msg["data"])
        target = payload.get("target")
        sender_pod = payload.get("pod")
        if sender_pod == POD_NAME:
            continue
        if target and target in clients:
            ws = clients[target]
            seq_counters[target] += 1
            seq = seq_counters[target]
            frame = {"seq": seq, "data": payload["data"]}
            message_buffers[target].append(frame)
            if len(message_buffers[target]) > BUFFER_SIZE:
                message_buffers[target] = message_buffers[target][-BUFFER_SIZE:]
            try:
                await ws.send_json(frame)
            except Exception:
                pass
        elif not target:
            for cid, ws in list(clients.items()):
                seq_counters[cid] += 1
                seq = seq_counters[cid]
                frame = {"seq": seq, "data": payload["data"]}
                message_buffers[cid].append(frame)
                if len(message_buffers[cid]) > BUFFER_SIZE:
                    message_buffers[cid] = message_buffers[cid][-BUFFER_SIZE:]
                try:
                    await ws.send_json(frame)
                except Exception:
                    pass


@asynccontextmanager
async def lifespan(app: FastAPI):
    global redis_pool, pubsub_task
    redis_pool = redis.from_url(REDIS_URL)
    pubsub_task = asyncio.create_task(backplane_listener())
    log.info("ws-server started pod=%s", POD_NAME)
    yield
    pubsub_task.cancel()
    await redis_pool.aclose()


app = FastAPI(title="WebSocket Server", lifespan=lifespan)
FastAPIInstrumentor.instrument_app(app)


@app.get("/healthz")
async def healthz():
    return {"status": "ok", "pod": POD_NAME}


@app.get("/info")
async def info():
    return {"pod": POD_NAME, "clients": list(clients.keys())}


@app.post("/send")
async def send_via_backplane(target: str = None, message: str = "hello"):
    payload = {"pod": POD_NAME, "target": target, "data": message}
    await redis_pool.publish(CHANNEL, json.dumps(payload))

    if target and target in clients:
        ws = clients[target]
        seq_counters[target] += 1
        seq = seq_counters[target]
        frame = {"seq": seq, "data": message}
        message_buffers[target].append(frame)
        if len(message_buffers[target]) > BUFFER_SIZE:
            message_buffers[target] = message_buffers[target][-BUFFER_SIZE:]
        await ws.send_json(frame)
    elif not target:
        for cid, ws in list(clients.items()):
            seq_counters[cid] += 1
            seq = seq_counters[cid]
            frame = {"seq": seq, "data": message}
            message_buffers[cid].append(frame)
            if len(message_buffers[cid]) > BUFFER_SIZE:
                message_buffers[cid] = message_buffers[cid][-BUFFER_SIZE:]
            try:
                await ws.send_json(frame)
            except Exception:
                pass

    return {"sent": True, "pod": POD_NAME}


@app.websocket("/ws/{client_id}")
async def websocket_endpoint(websocket: WebSocket, client_id: str):
    await websocket.accept()
    resume_seq = websocket.query_params.get("resume_seq")

    clients[client_id] = websocket
    log.info("client connected id=%s pod=%s", client_id, POD_NAME)

    if resume_seq:
        resume_seq = int(resume_seq)
        missed = [m for m in message_buffers.get(client_id, []) if m["seq"] > resume_seq]
        for frame in missed:
            await websocket.send_json(frame)
        log.info("replayed %d missed messages for client=%s from seq=%d", len(missed), client_id, resume_seq)

    try:
        while True:
            data = await websocket.receive_text()
            msg = json.loads(data) if data.startswith("{") else {"type": "ping"}
            if msg.get("type") == "ping":
                await websocket.send_json({"type": "pong", "pod": POD_NAME})
    except WebSocketDisconnect:
        log.info("client disconnected id=%s pod=%s", client_id, POD_NAME)
    finally:
        clients.pop(client_id, None)

File: examples/16-websockets/cpp/ws-server/src/main.cpp
#include <drogon/drogon.h>
#include <drogon/WebSocketController.h>
#include <sw/redis++/redis++.h>
#include <mutex>
#include <unordered_map>
#include <deque>
#include <thread>
#include <atomic>
#include <csignal>
#include <iostream>

using namespace drogon;

static std::string g_pod_name;
static std::string g_redis_url;
static std::unique_ptr<sw::redis::Redis> g_redis;
static std::atomic<bool> g_shutdown{false};

void signal_handler(int) { g_shutdown = true; }

struct ClientState {
    WebSocketConnectionPtr conn;
    int64_t seq = 0;
    std::deque<Json::Value> buffer;
    static constexpr size_t MAX_BUFFER = 100;
};

static std::mutex g_clients_mu;
static std::unordered_map<std::string, ClientState> g_clients;

static void send_frame(const std::string& client_id, const std::string& data) {
    std::lock_guard<std::mutex> lock(g_clients_mu);
    auto it = g_clients.find(client_id);
    if (it == g_clients.end()) return;

    auto& cs = it->second;
    cs.seq++;
    Json::Value frame;
    frame["seq"] = static_cast<Json::Int64>(cs.seq);
    frame["data"] = data;

    cs.buffer.push_back(frame);
    if (cs.buffer.size() > ClientState::MAX_BUFFER)
        cs.buffer.pop_front();

    Json::FastWriter writer;
    writer.omitEndingLineFeed();
    cs.conn->send(writer.write(frame));
}

static void broadcast_local(const std::string& data) {
    std::lock_guard<std::mutex> lock(g_clients_mu);
    for (auto& [cid, cs] : g_clients) {
        cs.seq++;
        Json::Value frame;
        frame["seq"] = static_cast<Json::Int64>(cs.seq);
        frame["data"] = data;

        cs.buffer.push_back(frame);
        if (cs.buffer.size() > ClientState::MAX_BUFFER)
            cs.buffer.pop_front();

        Json::FastWriter writer;
        writer.omitEndingLineFeed();
        cs.conn->send(writer.write(frame));
    }
}

static void redis_subscriber_loop() {
    try {
        auto sub = sw::redis::Subscriber(g_redis->subscriber());
        sub.subscribe("ws:broadcast");
        sub.on_message([](const std::string&, const std::string& msg) {
            Json::CharReaderBuilder builder;
            Json::Value payload;
            std::istringstream stream(msg);
            if (!Json::parseFromStream(builder, stream, &payload, nullptr)) return;

            std::string sender_pod = payload.get("pod", "").asString();
            if (sender_pod == g_pod_name) return;

            std::string target = payload.get("target", "").asString();
            std::string data = payload.get("data", "").asString();

            if (!target.empty()) {
                send_frame(target, data);
            } else {
                broadcast_local(data);
            }
        });

        while (!g_shutdown) {
            try {
                sub.consume();
            } catch (const sw::redis::TimeoutError&) {
                continue;
            }
        }
    } catch (const std::exception& e) {
        std::cerr << "Redis subscriber error: " << e.what() << std::endl;
    }
}

class WsCtrl : public WebSocketController<WsCtrl> {
public:
    void handleNewMessage(const WebSocketConnectionPtr& conn,
                          std::string&& message,
                          const WebSocketMessageType& type) override {
        if (type != WebSocketMessageType::Text) return;

        Json::CharReaderBuilder builder;
        Json::Value msg;
        std::istringstream stream(message);
        if (!Json::parseFromStream(builder, stream, &msg, nullptr)) return;

        if (msg.get("type", "").asString() == "ping") {
            Json::Value pong;
            pong["type"] = "pong";
            pong["pod"] = g_pod_name;
            Json::FastWriter writer;
            writer.omitEndingLineFeed();
            conn->send(writer.write(pong));
        }
    }

    void handleNewConnection(const HttpRequestPtr& req,
                             const WebSocketConnectionPtr& conn) override {
        std::string client_id = req->getHeader("X-Client-Id");
        if (client_id.empty()) client_id = "unknown";

        conn->setContext(std::make_shared<std::string>(client_id));

        {
            std::lock_guard<std::mutex> lock(g_clients_mu);
            g_clients[client_id] = {conn, 0, {}};
        }

        std::cerr << "client connected id=" << client_id << " pod=" << g_pod_name << std::endl;
    }

    void handleConnectionClosed(const WebSocketConnectionPtr& conn) override {
        auto ctx = conn->getContext<std::string>();
        if (!ctx) return;

        std::lock_guard<std::mutex> lock(g_clients_mu);
        g_clients.erase(*ctx);
        std::cerr << "client disconnected id=" << *ctx << " pod=" << g_pod_name << std::endl;
    }

    WS_PATH_LIST_BEGIN
    WS_PATH_ADD("/ws", Get);
    WS_PATH_LIST_END
};

int main() {
    std::signal(SIGTERM, signal_handler);
    std::signal(SIGINT, signal_handler);

    const char* env;

    env = std::getenv("POD_NAME");
    g_pod_name = env ? env : "ws-pod-unknown";

    env = std::getenv("REDIS_URL");
    g_redis_url = env ? env : "redis://redis:6379";

    sw::redis::ConnectionOptions opts;
    auto sep = g_redis_url.find("://");
    std::string host_port = (sep != std::string::npos) ? g_redis_url.substr(sep + 3) : g_redis_url;
    auto colon = host_port.find(':');
    opts.host = (colon != std::string::npos) ? host_port.substr(0, colon) : host_port;
    opts.port = (colon != std::string::npos) ? std::stoi(host_port.substr(colon + 1)) : 6379;
    opts.socket_timeout = std::chrono::milliseconds(1000);

    g_redis = std::make_unique<sw::redis::Redis>(opts);

    std::thread subscriber(redis_subscriber_loop);

    app().setLogLevel(trantor::Logger::kInfo);
    app().addListener("0.0.0.0", 8080);
    app().setUploadPath("/tmp");

    app().registerPreRoutingAdvice(
        [](const HttpRequestPtr& req,
           std::function<void(const HttpResponsePtr&)>&& callback,
           std::function<void()>&& pass) {
            auto path = req->path();
            if (path.size() > 4 && path.substr(0, 4) == "/ws/") {
                req->addHeader("X-Client-Id", path.substr(4));
                req->setPath("/ws");
            }
            pass();
        });

    app().registerHandler("/healthz",
        [](const HttpRequestPtr&, std::function<void(const HttpResponsePtr&)>&& callback) {
            Json::Value r;
            r["status"] = "ok";
            r["pod"] = g_pod_name;
            callback(HttpResponse::newHttpJsonResponse(r));
        },
        {Get});

    app().registerHandler("/info",
        [](const HttpRequestPtr&, std::function<void(const HttpResponsePtr&)>&& callback) {
            Json::Value r;
            r["pod"] = g_pod_name;
            Json::Value clients(Json::arrayValue);
            {
                std::lock_guard<std::mutex> lock(g_clients_mu);
                for (auto& [cid, _] : g_clients)
                    clients.append(cid);
            }
            r["clients"] = clients;
            callback(HttpResponse::newHttpJsonResponse(r));
        },
        {Get});

    app().registerHandler("/send",
        [](const HttpRequestPtr& req, std::function<void(const HttpResponsePtr&)>&& callback) {
            std::string target = req->getParameter("target");
            std::string message = req->getParameter("message");
            if (message.empty()) message = "hello";

            Json::Value payload;
            payload["pod"] = g_pod_name;
            payload["data"] = message;
            if (!target.empty()) payload["target"] = target;

            Json::FastWriter writer;
            writer.omitEndingLineFeed();
            std::string pub_msg = writer.write(payload);

            try {
                g_redis->publish("ws:broadcast", pub_msg);
            } catch (const std::exception& e) {
                std::cerr << "Redis publish error: " << e.what() << std::endl;
            }

            if (!target.empty()) {
                send_frame(target, message);
            } else {
                broadcast_local(message);
            }

            Json::Value r;
            r["sent"] = true;
            r["pod"] = g_pod_name;
            callback(HttpResponse::newHttpJsonResponse(r));
        },
        {Post});

    std::cerr << "ws-server started pod=" << g_pod_name << std::endl;
    app().run();

    g_shutdown = true;
    subscriber.join();
    return 0;
}
